lib/wasi/src/state/waker.rs

```diff
@@ -0,0 +1,59 @@
+use std::{
+    sync::Arc,
+    task::{RawWaker, RawWakerVTable, Waker},
+};
+use tokio::sync::mpsc;
+
+use wasmer_wasix_types::wasi::WakerId;
+
+use super::WasiState;
+
+#[derive(Clone)]
+pub(crate) struct WasiWaker {
+    id: WakerId,
+    tx: mpsc::UnboundedSender<(WakerId, bool)>,
+}
+impl WasiWaker {
+    pub(crate) fn new(id: WakerId, tx: &mpsc::UnboundedSender<(WakerId, bool)>) -> Self {
+        Self { id, tx: tx.clone() }
+    }
+
+    fn wake_now(&self) {
+        self.tx.send((self.id, true)).ok();
+    }
+}
+impl Drop for WasiWaker {
+    fn drop(&mut self) {
+        self.tx.send((self.id, false)).ok();
+    }
+}
+
+fn wasi_waker_wake(s: &WasiWaker) {
+    let waker_arc = unsafe { Arc::from_raw(s) };
+    waker_arc.wake_now();
+}
+
+fn wasi_waker_clone(s: &WasiWaker) -> RawWaker {
+    let arc = unsafe { Arc::from_raw(s) };
+    std::mem::forget(arc.clone());
+    RawWaker::new(Arc::into_raw(arc) as *const (), &VTABLE)
+}
+
+const VTABLE: RawWakerVTable = unsafe {
+    RawWakerVTable::new(
+        |s| wasi_waker_clone(&*(s as *const WasiWaker)), // clone
+        |s| wasi_waker_wake(&*(s as *const WasiWaker)),  // wake
+        |s| (*(s as *const WasiWaker)).wake_now(),       // wake by ref (don't decrease refcount)
+        |s| drop(Arc::from_raw(s as *const WasiWaker)),  // decrease refcount
+    )
+};
+
+fn wasi_waker_into_waker(s: *const WasiWaker) -> Waker {
+    let raw_waker = RawWaker::new(s as *const (), &VTABLE);
```

Comment:
Mmmm, are you creating an object "manualy" here? That looks a bit hacky.

Reply:
that's how `cooked_waker` does it, I was trying to avoid adding yet another dependency
https://crates.io/crates/cooked-waker

so i just copied the native implementation instead
https://medium.com/@KevinBGreene/async-programming-in-rust-part-3-futures-and-async-await-b508f7e44abc

it's the "right" way to do it apparently, its just ugly and rather not `rust` like